### PR TITLE
Allow more tile and wall conversion

### PR DESCRIPTION
**Pull request by sgkoishi**
Fixes unable to use Purification/Evil Powder in jungle, reported by [toxic](https://ptb.discord.com/channels/479657350043664384/479657350043664386/1046594200931667978).

### CODE REVIEW

**Comment by sgkoishi** [comment]
This actually add more allowed conversion other than jungle:
`MergesWithDirtInASpecialWay` set:
```
TileID.Grass
TileID.CorruptGrass
TileID.HallowedGrass
TileID.CrimsonGrass
TileID.GolfGrass
TileID.GolfGrassHallowed
```

`JungleGrass` set:
```
TileID.JungleGrass
TileID.CorruptJungleGrass
TileID.CrimsonJungleGrass
```

`GolfGrass` set:
```
TileID.GolfGrass
TileID.GolfGrassHallowed
```
<hr>
and walls

`Dirt` set:
```
WallID.DirtUnsafe
WallID.Dirt
```

`Snow` set:
```
WallID.SnowWallUnsafe
WallID.SnowWallEcho
```

`Ice` set:
```
WallID.IceUnsafe
WallID.IceEcho
```

File: TShockAPI/Handlers/SendTileRectHandler.cs [hunk 1]
```diff
@@ -398,36 +401,22 @@ internal void ProcessFlowerBoots(int realX, int realY, NetTile newTile)
...
-				TileID.Sets.Conversion.MossBrick[tile.type] && TileID.Sets.Conversion.MossBrick[newTile.Type]
-			)
+			foreach (var tileType in _convertibleTiles)
 			{
-				TShock.Log.ConsoleDebug(GetString($"Bouncer / SendTileRect processing a tile conversion update - [{tile.type}] -> [{newTile.Type}]"));
-				UpdateServerTileState(tile, newTile, TileDataType.Tile);
+				if (tileType[tile.type] && tileType[newTile.Type])
+				{
+					TShock.Log.ConsoleDebug(GetString($"Bouncer / SendTileRect processing a tile conversion update - [{tile.type}] -> [{newTile.Type}]"));
+					UpdateServerTileState(tile, newTile, TileDataType.Tile);
```
Comment (drunderscore) [inline TShockAPI/Handlers/SendTileRectHandler.cs]: Can we `break` early if this conversion is valid? I know the original code didn't, but it makes sense to do, and I believe it remains just fine and compatible with vanilla.

File: TShockAPI/Handlers/SendTileRectHandler.cs [hunk 2]
```diff
@@ -398,36 +401,22 @@ internal void ProcessFlowerBoots(int realX, int realY, NetTile newTile)
...
-				WallID.Sets.Conversion.NewWall4[tile.wall] && WallID.Sets.Conversion.NewWall4[newTile.Wall]
-			)
+			foreach (var wallType in _convertibleWalls)
 			{
-				TShock.Log.ConsoleDebug(GetString($"Bouncer / SendTileRect processing a wall conversion update - [{tile.wall}] -> [{newTile.Wall}]"));
-				UpdateServerTileState(tile, newTile, TileDataType.Wall);
+				if (wallType[tile.wall] && wallType[newTile.Wall])
+				{
+					TShock.Log.ConsoleDebug(GetString($"Bouncer / SendTileRect processing a wall conversion update - [{tile.wall}] -> [{newTile.Wall}]"));
+					UpdateServerTileState(tile, newTile, TileDataType.Wall);
```
Comment (drunderscore) [inline TShockAPI/Handlers/SendTileRectHandler.cs:447]: Ditto

**Comment by sgkoishi** [comment]
According to the client-side projectile AI `WorldGen.Convert(int i, int j, int conversionType, int size = 4)`, the `TileID.Sets.Conversion.Moss` and `TileID.Sets.Conversion.MossBrick` is never used by any vanilla code (perhaps we should remove them?)
(`Moss` is actually used in some checks like gemtree and shimmer, but neither is used in conversion)

**Review by drunderscore — CHANGES_REQUESTED** [review]
This does solve the issue of not being able to use Purification Powder on Crimson Jungle Grass. This implementation for sure seems to future-proof us much better!

**Comment by drunderscore** [comment]
@sgkoishi Yeah, we should only allow the conversions that vanilla allows, being as strict and accurate as possible. A filter on the field name should suffice?

**Review by hakusaro — DISMISSED** [review]

**Comment by hakusaro** [comment]
@drunderscore @sgkoishi hi! I didn't merge this because @drunderscore has a change request on this PR, fyi.

**Comment by sgkoishi** [comment]
Yes, it's still under consideration which impl is better. The proposed version fixed the problem and seems to be future-proof, but it actually allows certain impossible conversion (which those are harmless for now).

**Comment by sgkoishi** [comment]
The wall conversion is actually way more complicated - perhaps create dummy tiles and run `WorldGen.Convert` would be better to get an 'accurate' convertible table.

**Review by drunderscore — APPROVED** [review]
Code looks good, field filtering is also fine.
Tested and it does fix the issues talked about in Discord.

**Review by hakusaro — APPROVED** [review]

**Pull request merged**